In [1]:
!nvidia-smi

Fri Jun 24 06:18:50 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.32.03    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   49C    P8    10W /  70W |      0MiB / 15109MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
! pip install -q datasets

     |████████████████████████████████| 362 kB 5.2 MB/s 
     |████████████████████████████████| 140 kB 71.0 MB/s 
     |████████████████████████████████| 212 kB 56.8 MB/s 
     |████████████████████████████████| 1.1 MB 63.1 MB/s 
     |████████████████████████████████| 101 kB 14.8 MB/s 
     |████████████████████████████████| 596 kB 61.8 MB/s 
     |████████████████████████████████| 127 kB 75.5 MB/s 
     |████████████████████████████████| 271 kB 72.6 MB/s 
     |████████████████████████████████| 144 kB 78.8 MB/s 
     |████████████████████████████████| 94 kB 4.5 MB/s 
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
datascience 0.10.6 requires folium==0.2.1, but you have folium 0.8.3 which is incompatible.


In [4]:
! pip install transformers sentencepiece

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 4.4 MB 5.2 MB/s 
     |████████████████████████████████| 1.2 MB 52.3 MB/s 
     |████████████████████████████████| 6.6 MB 47.9 MB/s 


In [5]:
! pip install torch==1.6.0

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 748.8 MB 18 kB/s 
  Attempting uninstall: torch
    Found existing installation: torch 1.11.0+cu113
    Uninstalling torch-1.11.0+cu113:
      Successfully uninstalled torch-1.11.0+cu113
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
torchvision 0.12.0+cu113 requires torch==1.11.0, but you have torch 1.6.0 which is incompatible.
torchtext 0.12.0 requires torch==1.11.0, but you have torch 1.6.0 which is incompatible.
torchaudio 0.11.0+cu113 requires torch==1.11.0, but you have torch 1.6.0 which is incompatible.
fastai 2.6.3 requires torch<1.12,>=1.7.0, but you have torch 1.6.0 which is incompatible.


In [6]:
! pip install transformers

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [7]:
import pandas as pd
import numpy as np
import os
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/dataset/kaggle//us-patent/train.csv')
eval_df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/dataset/kaggle//us-patent/test.csv')

## Score meanings
The scores are in the 0-1 range with increments of 0.25 with the following meanings:
```
1.0 - Very close match. This is typically an exact match except possibly for differences in conjugation, quantity (e.g. singular vs. plural), and addition or removal of stopwords (e.g. “the”, “and”, “or”).
0.75 - Close synonym, e.g. “mobile phone” vs. “cellphone”. This also includes abbreviations, e.g. "TCP" -> "transmission control protocol".
0.5 - Synonyms which don’t have the same meaning (same function, same properties). This includes broad-narrow (hyponym) and narrow-broad (hypernym) matches.
0.25 - Somewhat related, e.g. the two phrases are in the same high level domain but are not synonyms. This also includes antonyms.
0.0 - Unrelated.
```

## Columns



*   id - a unique identifier for a pair of phrases
*   anchor - the first phrase
*   target - the second phrase
*   context - the CPC classification (version 2021.05), which indicates the subject 
*   within which the similarity is to be scored
*   score - the similarity. This is sourced from a combination of one or more 
manual expert ratings.





In [8]:
df

id        anchor                  target context  score
0      37d61fd2272659b1     abatement  abatement of pollution     A47   0.50
1      7b9652b17b68b7a4     abatement          act of abating     A47   0.75
2      36d72442aefd8232     abatement         active catalyst     A47   0.25
3      5296b0c19e1ce60e     abatement     eliminating process     A47   0.50
4      54c1e3b9184cb5b6     abatement           forest region     A47   0.00
...                 ...           ...                     ...     ...    ...
36468  8e1386cbefd7f245  wood article          wooden article     B44   1.00
36469  42d9e032d1cd3242  wood article              wooden box     B44   0.50
36470  208654ccb9e14fa3  wood article           wooden handle     B44   0.50
36471  756ec035e694722b  wood article         wooden material     B44   0.75
36472  8d135da0b55b8c88  wood article        wooden substrate     B44   0.50

[36473 rows x 5 columns]

In [9]:
eval_df

id                       anchor  \
0   4112d61851461f60                     opc drum   
1   09e418c93a776564              adjust gas flow   
2   36baf228038e314b               lower trunnion   
3   1f37ead645e7f0c8                cap component   
4   71a5b6ad068d531f           neural stimulation   
5   474c874d0c07bd21                     dry corn   
6   442c114ed5c4e3c9          tunneling capacitor   
7   b8ae62ea5e1d8bdb      angular contact bearing   
8   faaddaf8fcba8a3f  produce liquid hydrocarbons   
9   ae0262c02566d2ce             diesel fuel tank   
10  a8808e31641e856d            chemical activity   
11  16ae4b99d3601e60         transmit to platform   
12  25c555ca3d5a2092                  oil tankers   
13  5203a36c501f1b7c            generate in layer   
14  b9fdc772bb8fd61c                 slip segment   
15  7aa5908a77a7ec24                   el display   
16  d19ef3979396d47e              overflow device   
17  fd83613b7843f5e1     beam traveling direction   
18  2a619016908bfa45                   el display   
19  733979d75f59770d               equipment unit   
20  6546846df17f9800                   halocarbyl   
21  3ff0e7a35015be69         perfluoroalkyl group   
22  12ca31f018a2e2b9          speed control means   
23  03ba802ed4029e4d                   arm design   
24  c404f8b378cbb008               hybrid bearing   
25  78243984c02a72e4                     end pins   
26  de51114bc0faec3e             organic starting   
27  7e3aff857f056bf9                make of slabs   
28  26c3c6dc6174b589                   seal teeth   
29  b892011ab2e2cabc            carry by platform   
30  8247ff562ca185cc                        polls   
31  c057aecbba832387              upper clamp arm   
32  9f2279ce667b21dc              clocked storage   
33  b9ea2b06a878df6f              coupling factor   
34  79795133c30ef097       different conductivity   
35  25522ee5411e63e9               hybrid bearing   

                           target context  
0   inorganic photoconductor drum     G02  
1               altering gas flow     F23  
2                  lower locating     B60  
3                   upper portion     D06  
4       artificial neural network     H04  
5                 dry corn starch     C12  
6               capacitor housing     G11  
7       contact therapy radiation     B23  
8        produce a treated stream     C10  
9               diesel fuel tanks     F02  
10     dielectric characteristics     B01  
11               direct receiving     H04  
12                   oil carriers     B63  
13              generate by layer     G02  
14                   slip portion     B22  
15                   illumination     G02  
16                     oil filler     E04  
17                  concrete beam     H05  
18             electroluminescent     C23  
19                power detection     H02  
20      halogen addition reaction     C07  
21                        hydroxy     A63  
22                   control loop     G05  
23                    steel plate     F16  
24                 bearing system     F04  
25                       end days     A44  
26                organic farming     B61  
27                    making cake     E04  
28                teeth whitening     F01  
29              carry on platform     B60  
30                 pooling device     B21  
31                     end visual     A61  
32         clocked storage device     G01  
33                turns impedance     G01  
34               carrier polarity     H03  
35            corrosion resistant     F16

*include*：这个参数默认是只计算数值型特征的统计量，当输入include=['O']，会计算离散型变量的统计特征

In [10]:
df.describe(include='O')

id                       anchor       target context
count              36473                        36473        36473   36473
unique             36473                          733        29340     106
top     37d61fd2272659b1  component composite coating  composition     H01
freq                   1                          152           24    2186

We can see that in the 36473 rows, there are 733 unique anchors, 106 contexts, and nearly 30000 targets. Some anchors are very common, with "component composite coating" for instance appearing 152 times.

In [11]:
df['input'] = 'TEXT1:' + df.context + '; TEXT2:' + df.target + '; ANC1:' + df.anchor
df.head()

id     anchor                  target context  score  \
0  37d61fd2272659b1  abatement  abatement of pollution     A47   0.50   
1  7b9652b17b68b7a4  abatement          act of abating     A47   0.75   
2  36d72442aefd8232  abatement         active catalyst     A47   0.25   
3  5296b0c19e1ce60e  abatement     eliminating process     A47   0.50   
4  54c1e3b9184cb5b6  abatement           forest region     A47   0.00   

                                               input  
0  TEXT1:A47; TEXT2:abatement of pollution; ANC1:...  
1    TEXT1:A47; TEXT2:act of abating; ANC1:abatement  
2   TEXT1:A47; TEXT2:active catalyst; ANC1:abatement  
3  TEXT1:A47; TEXT2:eliminating process; ANC1:aba...  
4     TEXT1:A47; TEXT2:forest region; ANC1:abatement

## datasets Dataset类
### 构造方法
```
函数原型
datasets.Dataset(
		arrow_table: Table,
        info: Optional[DatasetInfo] = None,
        split: Optional[NamedSplit] = None,
        indices_table: Optional[Table] = None,
        fingerprint: Optional[str] = None,
    )
函数说明
根据ArrowTable产生一个Dataset对象。
```
**类属性/方法**\
1、from_dict函数\
根据字典来创建Dataset数据集对象。\
2、from_pandas函数\
根据pandas中的DataFrame对象创建Dataset对象。\
3、from_csv函数\
根据csv文件来创建Dataset对象。\
4、from_json函数\
5、from_text函数\
....

**对象属性/方法**
1、data属性\
获取Dataset中的数据。

2、cache_files属性\
获取Dataset中数据所在的缓存文件。

3、num_columns、num_rows属性\
获取Dataset中数据的列数、行数。

4、column_names属性\
获取Dataset中的列名称。

5、shape属性\
获取Dataset中的数据形状。

6、unique函数\
在指定列中返回一个不重复的列表。

7、add_column函数\
在Dataset数据集中增加一列数据。

8、add_item函数\
在Dataset数据集中增加一行数据。

9、cast函数\
将Dataset中特征features装换成新的features。

10、cast_column函数\
转换features中的某一列。


## DatasetDict
Dictionary with split names as keys (‘train’, ‘test’ for example), and Dataset objects as values. It also has dataset transform methods like map or filter, to process all the splits at once.

以拆分名称作为键（例如“train”、“test”）和 Dataset 对象作为值的字典。它还具有数据集转换方法，例如 map 或 filter，可以一次处理所有拆分。

### map
Apply a function to all the elements in the table (individually or in batches) and update the table (if function does updated examples). The transformation is applied to all the datasets of the dataset dictionary.

将函数应用于表中的所有元素（单独或批量）并更新表（如果函数确实更新了示例）。转换应用于数据集字典的所有数据集。
```
**function** (callable) — with one of the following signature:
  function(example: Dict[str, Any]) -> Dict[str, Any] if batched=False and with_indices=False
  function(example: Dict[str, Any], indices: int) -> Dict[str, Any] if batched=False and with_indices=True
  function(batch: Dict[str, List]) -> Dict[str, List] if batched=True and with_indices=False
  function(batch: Dict[str, List], indices: List[int]) -> Dict[str, List] if batched=True and with_indices=True

For advanced usage, the function can also return a pyarrow.Table. Moreover if your function returns nothing (None), then map will run your function and return the dataset unchanged.

**batched** (bool, defaults to False) — Provide batch of examples to function
```

In [12]:
from datasets import Dataset,DatasetDict
ds = Dataset.from_pandas(df)
ds

Dataset({
    features: ['id', 'anchor', 'target', 'context', 'score', 'input'],
    num_rows: 36473
})

In [13]:
from transformers import BertTokenizer
from transformers import AutoModelForSequenceClassification,AutoTokenizer
import torch
# Bert_model = '/content/drive/MyDrive/Colab Notebooks/dataset/BERT_model'
deberta_model = 'microsoft/deberta-v3-small'
tokenizer = AutoTokenizer.from_pretrained(deberta_model)

Downloading:   0%|          | 0.00/52.0 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/578 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/2.35M [00:00<?, ?B/s]

Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
/usr/local/lib/python3.7/dist-packages/transformers/convert_slow_tokenizer.py:435: UserWarning: The sentencepiece tokenizer that you are converting to a fast tokenizer uses the byte fallback option which is not implemented in the fast tokenizers. In practice this means that the fast version of the tokenizer can produce unknown tokens whereas the sentencepiece version would have converted these unknown tokens into a sequence of byte tokens matching the original piece of text.
  "The sentencepiece tokenizer that you are converting to a fast tokenizer uses the byte fallback option"
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.


In [14]:
tokenizer.tokenize("A platypus is an ornithorhynchus anatinus.")

['▁A',
 '▁platypus',
 '▁is',
 '▁an',
 '▁or',
 'ni',
 'tho',
 'rhynch',
 'us',
 '▁an',
 'at',
 'inus',
 '.']

In [15]:
tokenizer("A platypus is an ornithorhynchus anatinus.")

{'input_ids': [1, 336, 114224, 269, 299, 289, 4840, 34765, 102530, 1867, 299, 2401, 26835, 260, 2], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}

In [16]:
def tok_func(x): return tokenizer(x['input'])

In [17]:
tok_ds = ds.map(tok_func, batched = True)

Parameter 'function'=<function tok_func at 0x7f9991cb5b00> of the transform datasets.arrow_dataset.Dataset._map_single couldn't be hashed properly, a random hash was used instead. Make sure your transforms and parameters are serializable with pickle or dill for the dataset fingerprinting and caching to work. If you reuse this transform, the caching mechanism will consider it to be different from the previous calls and recompute everything. This warning is only showed once. Subsequent hashing failures won't be showed.


  0%|          | 0/37 [00:00<?, ?ba/s]

In [18]:
tok_ds

Dataset({
    features: ['id', 'anchor', 'target', 'context', 'score', 'input', 'input_ids', 'token_type_ids', 'attention_mask'],
    num_rows: 36473
})

In [19]:
tok_ds[0]

{'anchor': 'abatement',
 'attention_mask': [1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1],
 'context': 'A47',
 'id': '37d61fd2272659b1',
 'input': 'TEXT1:A47; TEXT2:abatement of pollution; ANC1:abatement',
 'input_ids': [1,
  54453,
  435,
  294,
  558,
  5753,
  346,
  54453,
  445,
  294,
  16191,
  297,
  17019,
  265,
  6435,
  346,
  23702,
  435,
  294,
  16191,
  297,
  17019,
  2],
 'score': 0.5,
 'target': 'abatement of pollution',
 'token_type_ids': [0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0]}

Okay, we need to prepare our labels. Transformers always assumes that your labels has the column name labels, but in our dataset it's currently score. Therefore, we need to rename it:

In [20]:
tok_ds = tok_ds.rename_columns({'score':'labels'})

Now that we've prepared our tokens and labels, we need to create our validation set.

## Test and validation sets

In [21]:
eval_df.describe()

id      anchor                         target context
count                 36          36                             36      36
unique                36          34                             36      29
top     4112d61851461f60  el display  inorganic photoconductor drum     G02
freq                   1           2                              1       3

This is the test set. Possibly the most important idea in machine learning is that of having separate training, validation, and test data sets.

The validation set is only ever used to see how we're doing. It's never used as inputs to training the model.

Transformers uses a DatasetDict for holding your training and validation sets. To create one that contains 25% of our data for the validation set, and 75% for the training set, use train_test_split:

In [22]:
from numpy.random import normal,seed,uniform
np.random.seed(2022)
train_and_val = tok_ds.train_test_split(0.25, seed = 2022)
train_and_val

DatasetDict({
    train: Dataset({
        features: ['id', 'anchor', 'target', 'context', 'labels', 'input', 'input_ids', 'token_type_ids', 'attention_mask'],
        num_rows: 27354
    })
    test: Dataset({
        features: ['id', 'anchor', 'target', 'context', 'labels', 'input', 'input_ids', 'token_type_ids', 'attention_mask'],
        num_rows: 9119
    })
})

As you see above, the validation set here is called test and not validate

In practice, a random split like we've used here might not be a good idea -- here's what Dr Rachel Thomas has to say about it:
```
"One of the most likely culprits for this disconnect between results in development vs results in production is a poorly chosen validation set (or even worse, no validation set at all). Depending on the nature of your data, choosing a validation set can be the most important step. Although sklearn offers a train_test_split method, this method takes a random subset of the data, which is a poor choice for many real-world problems."
开发结果与生产结果之间的这种脱节最可能的罪魁祸首之一是选择不当的验证集（或者更糟糕的是，根本没有验证集）。 根据数据的性质，选择验证集可能是最重要的一步。 尽管 sklearn 提供了一个 train_test_split 方法，但该方法采用数据的随机子集，这对于许多现实世界的问题来说是一个糟糕的选择。
```

In [23]:
eval_df['input'] = 'TEXT1:' + df.context + '; TEXT2:' + df.target + '; ANC1:' + df.anchor
eval_ds = Dataset.from_pandas(eval_df).map(tok_func, batched = True)

  0%|          | 0/1 [00:00<?, ?ba/s]

In [24]:
eval_ds

Dataset({
    features: ['id', 'anchor', 'target', 'context', 'input', 'input_ids', 'token_type_ids', 'attention_mask'],
    num_rows: 36
})

## Training

In [25]:
def corr(x,y): return np.corrcoef(x,y)[0][1]
def corr_d(eval_pred): return {'pearson': corr(*eval_pred)}

## Trainer
The Trainer class provides an API for feature-complete training in PyTorch for most standard use cases. It’s used in most of the example scripts.\
Before instantiating your Trainer, create a TrainingArguments to access all the points of customization during training.\
The API supports distributed training on multiple GPUs/TPUs, mixed precision through NVIDIA Apex and Native AMP for PyTorch.
```
The Trainer class is optimized for 🤗 Transformers models and can have surprising behaviors when you use it on other models. When using it on your own model, make sure:

-- your model always return tuples or subclasses of ModelOutput.
-- ☆☆☆your model can compute the loss if a labels argument is provided and that loss is returned as the first element of the tuple (if your model returns tuples)
-- your model can accept multiple label arguments (use the label_names in your TrainingArguments to indicate their name to the Trainer) but none of them should be named "label".
```

Here is an example of how to customize Trainer to use a weighted loss (useful when you have an unbalanced training set):
```
from torch import nn
from transformers import Trainer


class CustomTrainer(Trainer):
    def compute_loss(self, model, inputs, return_outputs=False):
        labels = inputs.get("labels")
        # forward pass
        outputs = model(**inputs)
        logits = outputs.get("logits")
        # compute custom loss (suppose one has 3 labels with different weights)
        loss_fct = nn.CrossEntropyLoss(weight=torch.tensor([1.0, 2.0, 3.0]))
        loss = loss_fct(logits.view(-1, self.model.config.num_labels), labels.view(-1))
        return (loss, outputs) if return_outputs else loss
```


Ref : https://huggingface.co/docs/transformers/main/en/main_classes/trainer#transformers.Trainer

In [26]:
from transformers import TrainingArguments, Trainer, AutoModelForSequenceClassification
bs = 128
epochs = 8
lr = 8e-5

args = TrainingArguments('outputs', 
              learning_rate = lr, 
              warmup_ratio = 0.1,
              lr_scheduler_type = 'cosine', 
              fp16 = True, 
              evaluation_strategy = 'epoch', 
              per_device_train_batch_size = bs,
              per_device_eval_batch_size = bs * 2, 
              num_train_epochs = epochs,
              weight_decay = 0.01, 
              report_to = 'none')
# report_to (str or List[str], optional, defaults to "all") 
# — The list of integrations to report the results and logs to. Supported platforms are "azure_ml", "comet_ml", "mlflow", "tensorboard" and "wandb". Use "all" to report to all integrations installed, "none" for no integrations.
# 要向其报告结果和日志的集成列表。 支持的平台是“azure_ml”、“comet_ml”、“mlflow”、“tensorboard”和“wandb”。 使用“all”报告所有已安装的集成，“none”表示没有集成。

model = AutoModelForSequenceClassification.from_pretrained(deberta_model, num_labels = 1)
trainer = Trainer(model, args, train_dataset = train_and_val['train'],
          eval_dataset = train_and_val['test'], tokenizer = tokenizer, 
          compute_metrics = corr_d)

Downloading:   0%|          | 0.00/273M [00:00<?, ?B/s]

Some weights of the model checkpoint at microsoft/deberta-v3-small were not used when initializing DebertaV2ForSequenceClassification: ['mask_predictions.classifier.weight', 'lm_predictions.lm_head.dense.bias', 'mask_predictions.LayerNorm.bias', 'mask_predictions.dense.weight', 'mask_predictions.LayerNorm.weight', 'mask_predictions.classifier.bias', 'lm_predictions.lm_head.dense.weight', 'lm_predictions.lm_head.bias', 'mask_predictions.dense.bias', 'lm_predictions.lm_head.LayerNorm.weight', 'lm_predictions.lm_head.LayerNorm.bias']
- This IS expected if you are initializing DebertaV2ForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing DebertaV2ForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from 

In [27]:
trainer.train();

The following columns in the training set don't have a corresponding argument in `DebertaV2ForSequenceClassification.forward` and have been ignored: target, context, input, anchor, id. If target, context, input, anchor, id are not expected by `DebertaV2ForSequenceClassification.forward`,  you can safely ignore this message.
/usr/local/lib/python3.7/dist-packages/transformers/optimization.py:310: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  FutureWarning,
***** Running training *****
  Num examples = 27354
  Num Epochs = 8
  Instantaneous batch size per device = 128
  Total train batch size (w. parallel, distributed & accumulation) = 128
  Gradient Accumulation steps = 1
  Total optimization steps = 1712


The following columns in the evaluation set don't have a corresponding argument in `DebertaV2ForSequenceClassification.forward` and have been ignored: target, context, input, anchor, id. If target, context, input, anchor, id are not expected by `DebertaV2ForSequenceClassification.forward`,  you can safely ignore this message.
***** Running Evaluation *****
  Num examples = 9119
  Batch size = 256
The following columns in the evaluation set don't have a corresponding argument in `DebertaV2ForSequenceClassification.forward` and have been ignored: target, context, input, anchor, id. If target, context, input, anchor, id are not expected by `DebertaV2ForSequenceClassification.forward`,  you can safely ignore this message.
***** Running Evaluation *****
  Num examples = 9119
  Batch size = 256
Saving model checkpoint to outputs/checkpoint-500
Configuration saved in outputs/checkpoint-500/config.json
Model weights saved in outputs/checkpoint-500/pytorch_model.bin
tokenizer config file saved 

In [28]:
prediction = trainer.predict(eval_ds).predictions.astype(float)
prediction

The following columns in the test set don't have a corresponding argument in `DebertaV2ForSequenceClassification.forward` and have been ignored: target, context, input, anchor, id. If target, context, input, anchor, id are not expected by `DebertaV2ForSequenceClassification.forward`,  you can safely ignore this message.
***** Running Prediction *****
  Num examples = 36
  Batch size = 256


array([[ 5.25878906e-01],
       [ 7.66113281e-01],
       [ 2.79052734e-01],
       [ 5.56640625e-01],
       [-2.71911621e-02],
       [ 2.25097656e-01],
       [ 4.80041504e-02],
       [ 9.08813477e-02],
       [ 5.84472656e-01],
       [ 2.30834961e-01],
       [ 5.76171875e-01],
       [ 5.13671875e-01],
       [ 5.19531250e-01],
       [ 4.89013672e-01],
       [-2.21710205e-02],
       [-2.46276855e-02],
       [ 2.26806641e-01],
       [ 5.34179688e-01],
       [-2.09503174e-02],
       [-1.77001953e-02],
       [ 1.07849121e-01],
       [ 5.19531250e-01],
       [ 5.35156250e-01],
       [ 2.47192383e-01],
       [ 5.19042969e-01],
       [ 5.26367188e-01],
       [ 1.54800415e-02],
       [ 5.26855469e-01],
       [ 1.05371094e+00],
       [-6.54220581e-04],
       [-7.49588013e-03],
       [ 2.00317383e-01],
       [ 1.52221680e-01],
       [-2.94952393e-02],
       [ 4.84130859e-01],
       [ 4.95117188e-01]])

Look out - some of our predictions are <0, or >1! This once again shows the value of remember to actually look at your data. Let's fix those out-of-bounds predictions:

In [29]:
pred = np.clip(prediction, 0, 1)
pred

array([[0.52587891],
       [0.76611328],
       [0.27905273],
       [0.55664062],
       [0.        ],
       [0.22509766],
       [0.04800415],
       [0.09088135],
       [0.58447266],
       [0.23083496],
       [0.57617188],
       [0.51367188],
       [0.51953125],
       [0.48901367],
       [0.        ],
       [0.        ],
       [0.22680664],
       [0.53417969],
       [0.        ],
       [0.        ],
       [0.10784912],
       [0.51953125],
       [0.53515625],
       [0.24719238],
       [0.51904297],
       [0.52636719],
       [0.01548004],
       [0.52685547],
       [1.        ],
       [0.        ],
       [0.        ],
       [0.20031738],
       [0.15222168],
       [0.        ],
       [0.48413086],
       [0.49511719]])

In [30]:
import datasets
submission = datasets.Dataset.from_dict(
    {
       'id': eval_ds['id'],
       'score': pred 
    }
)
submission

Dataset({
    features: ['id', 'score'],
    num_rows: 36
})

In [31]:
submission.to_csv('deberta_v3_small.csv', index = False)

Creating CSV from Arrow format:   0%|          | 0/1 [00:00<?, ?ba/s]

1017